bitswap/provider.go

```diff
@@ -94,264 +120,305 @@ func NewProvider(
+		retrievalQueue: peertaskqueue.New(peertaskqueue.TaskMerger(&overwriteTaskMerger{})),
 	}
 
 	provider.network.Start(provider)
 
-	for i := uint(0); i < config.MaxBitswapWorkers; i++ {
-		go provider.runWorker()
+	for i := 0; i < 8; i++ {
+		go provider.handleRequests()
+	}
+
+	for i := 0; i < 8; i++ {
+		go provider.handleResponses()
+	}
+
+	for i := 0; i < 8; i++ {
+		go provider.handleRetrievals()
 	}
 
 	return provider, nil
 }
 
-// Upon receiving a message, provider will iterate over the requested CIDs. For
-// each CID, it'll check if it's present in the blockstore. If it is, it will
-// respond with that block, and if it isn't, it'll start a retrieval and
-// register the block to be sent later.
 func (provider *Provider) ReceiveMessage(ctx context.Context, sender peer.ID, incoming message.BitSwapMessage) {
+	var tasks []peertask.Task
 	for _, entry := range incoming.Wantlist() {
+		tasks = append(tasks, peertask.Task{
+			Topic:    entry.Cid,
+			Priority: int(entry.Priority),
+			Work:     1,
+			Data:     entry,
+		})
+	}
+	provider.requestQueue.PushTasks(sender, tasks...)
+}
 
-		// Immediately, if this is a cancel message, just ignore it (TODO)
-		if entry.Cancel {
+func (provider *Provider) handleRequests() {
+	ctx := context.Background()
+
+	for {
+		peerID, tasks, _ := provider.requestQueue.PopTasks(100)
+		if len(tasks) == 0 {
+			time.Sleep(time.Millisecond * 250)
 			continue
 		}
 
-		stats.Record(ctx, metrics.BitswapRequestCount.M(1))
+		log.Debugf("Processing %d requests for %s", len(tasks), peerID)
 
-		// We want to skip CIDs in the blacklist, queue DONT_HAVE
-		if provider.config.CidBlacklist[entry.Cid] {
-			logger.Debugf("Replying DONT_HAVE for blacklisted CID: %s", entry.Cid)
-			provider.queueDontHave(ctx, sender, entry, "blacklisted_cid")
-			continue
+		for _, task := range tasks {
+			entry, ok := task.Data.(message.Entry)
+			if !ok {
+				log.Warnf("Invalid request entry data")
+				continue
+			}
+
+			provider.handleRequest(ctx, peerID, entry)
 		}
 
+		provider.requestQueue.TasksDone(peerID, tasks...)
+	}
+}
+
+func (provider *Provider) handleRequest(
+	ctx context.Context,
+	peerID peer.ID,
+	entry message.Entry,
+) error {
+	log := log.With("peer_id", peerID)
+
+	// Skip blacklisted CIDs
+	if provider.config.CidBlacklist[entry.Cid] {
+		log.Debugf("Replying DONT_HAVE for blacklisted CID: %s", entry.Cid)
+		provider.queueSendDontHave(peerID, int(entry.Priority), entry.Cid, "blacklisted_cid")
+		return nil
+	}
+
+	// If it's a cancel, just remove from the queue and finish
+
+	if entry.Cancel {
+		log.Debugf("Cancelling request for %s", entry.Cid)
+		provider.responseQueue.Remove(entry.Cid, peerID)
+		provider.retrievalQueue.Remove(entry.Cid, peerID)
+		return nil
+	}
+
+	stats.Record(ctx, metrics.BitswapRequestCount.M(1))
+
+	// First check blockstore, immediately write to response queue if it
+	// exists
+
+	size, err := provider.blockManager.GetSize(ctx, entry.Cid)
+	if err == nil {
```

Comment:
may be worth bringing back the not-not-found error logging for this as an `} else !ipld.IsNotFound(err) { ... logger.Warnf... }`

Reply:
added